Implmentation of contextual launch

jackiehanyang opened the pull request:
### Description
This PR implements the contextual launch feature. It first checks whether an index pattern with a title matching the indices used to create the detector already exists. If it does, the index pattern id is retrieved and included in the Discover page URL. If not, a new index pattern is created using all the indices that were used to create the detector, and its id is then included in the Discover page URL.

![Screenshot 2025-04-02 at 16 30 36](https://github.com/user-attachments/assets/4d067bf9-998d-4cf2-bd86-6165c35b0602)
![Screenshot 2025-04-02 at 16 30 55](https://github.com/user-attachments/assets/6ad36ed4-bc5b-4ea3-94e1-f8b72e30965c)

### Issues Resolved

[List any issues this PR will resolve]

### Check List

- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

kaituo commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,85 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+      // put query params for HC detector
+      let queryParams = '';
+      if (props.isHCDetector && item[ENTITY_VALUE_FIELD]) {
+        const [field, value] = item[ENTITY_VALUE_FIELD].split(': ').map((s: string) => s.trim());
```
This does not work for multiple categorical fields.

kaituo replied:
When I clicking host:server_1, service:app_3, this is what I get 

![Screenshot 2025-04-03 at 10 06 02 AM](https://github.com/user-attachments/assets/0f7875d5-3e3d-4365-9c57-6c0b796cf0d2)

jackiehanyang replied:
oops, forgot entity could have multiple values. 

updated the code logic to handle it and here's a screenshot in Discover that selects two entities.

also, in UTs, I'm using two entities values in the mocked anomalies data. 
![Screenshot 2025-04-03 at 14 37 17](https://github.com/user-attachments/assets/985dee35-0253-48df-ae47-d24b645e3b3e)

kaituo reviewed (commented):
Can you write some ITs or UTs to cover your functionality?

jackiehanyang commented on public/pages/DetectorResults/utils/tableUtils.tsx:
```diff
@@ -74,6 +74,29 @@ export const staticColumn = [
+      <EuiText size="xs" style={columnStyle}>
+        <b>Actions</b>{' '}
+        <EuiIconTip
+          content="This will create an index pattern with the indices used to create this detector and open the anomaly logs in Discover."
```
updated the text as `This will create an index pattern with the indices used to create this detector (if it doesn't exist) and open the anomaly logs in Discover.`

jackiehanyang commented:
> Can you write some ITs or UTs to cover your functionality?

Added a UT file. There was no existing UT setup for the file I touched, so I originally planned to cover it with an IT. However, after reviewing it, I felt it would be better to have UTs for this feature. So I set up a new file to add unit tests for the AnomalyResultsTable page.

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,99 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+          
+          indexPatternId = newIndexPattern.id;
+
+          getNotifications().toasts.addSuccess(`Created new index pattern: ${indexPatternTitle}`);
```
suggestion: utilize ToastInputFields https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/notifications/toasts/toasts_api.tsx#L48 to pass it it a title and id. 

this will ensure you don't have duplicate toasts on screen

jackiehanyang replied:
thanks for the suggestion, opened an issue about refactoring this cross the repo - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/issues/1006

kavilla commented on public/pages/DetectorResults/utils/tableUtils.tsx:
```diff
@@ -74,6 +74,30 @@ export const staticColumn = [
+        description: 'View in Discover',
+        icon: 'editorLink',
+        onClick: () => {},
+        'data-test-subj': 'discoverIcon',
```
apologize i dont have the insight on if this ends up getting info prepended to this test subj but the suggestion would be to make this specific to this page. to avoid potential conflicts. if tmrw the navigation bar adds a button with the name `discoverIcon` then cypress might find the wrong test subj

jackiehanyang replied:
I agree we should make test subj unique enough, but we should be fine here :)

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,99 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+        queryParams = `filters:!(${filters.join(',')}),`;
+      }
+
+      const discoverUrl = `${basePath}/app/data-explorer/discover#?_a=(discover:(columns:!(_source),isDirty:!f,sort:!()),metadata:(indexPattern:'${indexPatternId}',view:discover))&_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'${startISO}',to:'${endISO}'))&_q=(${queryParams}query:(language:kuery,query:''))`;
```
instead of this. you might be able to do something like
```ts
// destructure the query string manager 
 const {
        filterManager,
        queryString,
        timefilter: { timefilter },
      } = services.data.query;
      

const { history } = services;

// sync initial app state from state container to managers
filterManager.setAppFilters(filters);
queryString.setQuery({
  language: kuery,
  query: '',
  ${ifEnhancementsEnabled via advanced settings} && dataset
});
```

when enhancements enabled (so essentially workspaces and datasources enabled). we rip out the index pattern from there and shove it in the dataset within the query

more details:
 https://opensearch-project.github.io/OpenSearch-Dashboards/docs/index.html#/plugins/discover/understand_and_extend_discover?id=_5-the-query-system-and-flow

jackiehanyang replied:
Thanks for the suggestion! I looked into this approach, but I’d prefer to keep the current implementation as is. It’s working well and handles all the required scenarios, including when enhancements are enabled. The main feature requirement here is to help customers open the Discover page with some detector anomaly information pre-filled. Introducing the singleton-based query string manager approach would add extra complexity without clear benefits in this case. Let me know if there's a specific concern with the current implementation that I might have missed—happy to chat!

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -29,12 +28,18 @@ import { AnomalyData } from '../../../models/interfaces';
   isHCDetector?: boolean;
   isHistorical?: boolean;
   selectedHeatmapCell?: HeatmapCell | undefined;
+  detectorIndex: string[];
```
nitpick: should this be like detectorIndices? to match other fields

could potentially utilize the `dataset` type or `data structure` type https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/data/common/datasets/types.ts#L40 so that instead of passing two new props you just pass a dataset that has insight on the underlying indices and the timefieldName

jackiehanyang replied:
updated the prop name s `detectorIndices`

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,99 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+  const handleOpenDiscover = async (startTime: number, endTime: number, item: any) => {
+    try {
+      // calculate time range with 10-minute buffer on each side per customer request
+      const TEN_MINUTES_IN_MS = 10 * 60 * 1000;
```
nit: would suggest to export a const like this in a common folder. ex https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/telemetry/common/constants.ts#L53

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,99 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+      const startISO = new Date(startTime - TEN_MINUTES_IN_MS).toISOString();
+      const endISO = new Date(endTime + TEN_MINUTES_IN_MS).toISOString();
+      
+      const basePath = `${window.location.origin}${window.location.pathname.split('/app/')[0]}`;
```
instead you could use the core service that is accessible by every plugin `getUrlForApp` 
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/application/types.ts#L826

so like
```typescript
window.open(getUrlForApp(def.appId), '_blank', 'noopener, noreferrer')
```

```

jackiehanyang replied:
Thanks for the suggestion, I didn't know we have such method :) 
I would leave it as the current implementation because these two methods work exactly the same, but the one that uses the core service adds extra burden when writing unit tests, as it needs to be mocked.

kavilla replied:
if you plan on mock the services to be super basic, you should be able to use the helper methods which should instantiate all the mocks for those services

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/mocks.ts#L200

for example,
```typescript
import { coreMock } from '../../../../../core/public/mocks';

const startMock = coreMock.createStart();
```

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,99 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+    }
+  };
+
+  const getCustomColumns = () => {
```
not sure what these are used for so please ignore if this doesn't make sense. but you can see in the discover url there is a columns too. so technically can set the columns as well. if we wwanted to carry over some state information. but like i said if this a different type of columns then pls ignore

jackiehanyang replied:
Yeah it has nothing to do with the discover url. Different detectors on different pages will have different columns in this table. The column customization logic used to be scattered in various places, but in this PR, I consolidated all of it into this method :)

kavilla commented on public/pages/DetectorResults/containers/AnomalyResultsTable.tsx:
```diff
@@ -62,6 +67,99 @@ export function AnomalyResultsTable(props: AnomalyResultsTableProps) {
+      
+      window.open(discoverUrl, '_blank');
+    } catch (error) {
+      getNotifications().toasts.addDanger('Error opening discover view');
```
nit: add i18n

jackiehanyang replied:
added in the latest revision

kaituo commented:
Test failed:

Test Suites: 1 failed, 1 skipped, 85 passed, 86 of 87 total
Tests:       1 failed, 2 skipped, 441 passed, 444 total
Snapshots:   80 passed, 80 total
Time:        163.372 s
Ran all test suites.

jackiehanyang commented:
> Test failed:
> 
> Test Suites: 1 failed, 1 skipped, 85 passed, 86 of 87 total Tests: 1 failed, 2 skipped, 441 passed, 444 total Snapshots: 80 passed, 80 total Time: 163.372 s Ran all test suites.

The failed test case is flaky, so I deleted it in the latest revision. I tried to capture the error message in the inner try catch block but sometimes the error message from the outer try catch block is being captured, so it's flaky.

kaituo commented:
I started a detector with two categorical field with one local index and one remote index. When clicking the contextual link, I got
![Screenshot 2025-04-04 at 1 33 57 PM](https://github.com/user-attachments/assets/e1e0ae80-f700-42c4-bbba-febadae2200a)

please fix the error and add the scenario to your test.


Error getting aggregated anomaly results for detector Yxt_ApYBFppsB-v16WKP:  [1:155] [bool] failed to parse field [filter] 2 [AnomalyDetailsChart.tsx:115](http://localhost:5601/plugin:anomalyDetectionDashboards/plugins/anomaly-detection-dashboards-plugin/public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx)
    AnomalyDetailsChart AnomalyDetailsChart.tsx:115
Warning: Each child in a list should have a unique "key" prop. See https://fb.me/react-warning-keys for more information.
    in SpecsParserComponent (created by Connect(SpecsParserComponent))
    in Connect(SpecsParserComponent) (created by Chart)
    in ErrorBoundary (created by Chart)
    in div (created by Chart)
    in Provider (created by Chart)
    in Chart (created by FeatureChart)
    in div (created by FeatureChart)
    in div (created by ContentPanel)
    in div (created by ContentPanel)
    in div (created by EuiPanel)
    in EuiPanel (created by ContentPanel)
    in ContentPanel (created by FeatureChart)
    in FeatureChart
    in Unknown (created by AnomalyHistory)
    in div (created by EuiPanel)
    in EuiPanel (created by AnomalyHistory)
    in div (created by AnomalyHistory)
    in div (created by AnomalyHistory)
    in div
    in div (created by ContentPanel)
    in div (created by ContentPanel)
    in div (created by EuiPanel)
    in EuiPanel (created by ContentPanel)
    in ContentPanel
    in Unknown (created by AnomalyHistory)
    in AnomalyHistory (created by HistoricalDetectorResults)
    in div (created by EuiFlexItem)
    in EuiFlexItem (created by HistoricalDetectorResults)
    in div (created by EuiFlexGroup)
    in EuiFlexGroup (created by HistoricalDetectorResults)
    in div (created by EuiPageBody)
    in EuiPageBody (created by HistoricalDetectorResults)
    in div (created by EuiPage)
    in EuiPage (created by HistoricalDetectorResults)
    in HistoricalDetectorResults (created by Context.Consumer)
    in Route (created by DetectorDetail)
    in Switch (created by DetectorDetail)
    in DetectorDetail (created by Context.Consumer)
    in Route (created by Context.Consumer)
    in Switch (created by Context.Consumer)
    in div (created by EuiPageBody)
    in EuiPageBody (created by Context.Consumer)
    in div (created by EuiPage)
    in EuiPage (created by Context.Consumer)
    in Main (created by Context.Consumer)
    in Route
    in Router (created by HashRouter)
    in HashRouter
    in Provider [react-dom.development.js:88](http://localhost:5601/9007199254740991/bundles/osd-ui-shared-deps/osd-ui-shared-deps/node_modules/react-dom/cjs/react-dom.development.js)

​

jackiehanyang commented:
> AnomalyDetailsChart

the error looks like related to AnomalyDetailsChart when rendering the chart as some field value is off instead of contextual launch. But let me look into it.

jackiehanyang commented:
> > AnomalyDetailsChart
> 
> the error looks like related to AnomalyDetailsChart when rendering the chart as some field value is off instead of contextual launch. But let me look into it.

fixed in the new revision, the current implementation works with remote index use case, just needed also update the props name in the parent file when changing the props name. Fixed in the latest revision and added a test case for it
